### ISSUE TITLE
Mint clients should detect and resolve internal payments without hopping through a Gateway

### ISSUE THREAD

**Issue by okjodom** [fedimint/fedimint#486]
When a user client funds and pays for an outgoing LN invoice, there's possibility that the gateway is just contracted to route the payment for another user client within the same Mint as the first user. Instead of this inefficient hop over the LN network as facilitated by the Gateway, could the user clients directly infer an internal payment and settle it within the mint without going through the gateway?

Helps avoid the bug described in #461

**Comment by okjodom** [comment]
On my first attempt to tackle this issue, inferring internal payments from the routing hint is turning out to be hard, and I suspect it's because the routing hint as we have now may not be sufficient information.

Here is my first attempt at adding user client logic to infer internal payments from routing hint:
```rs
    /// Infer if the invoice presented is an internal payment within the mint
    ///
    /// We do this by checking if the last hop for the invoice is to a gateway that is registerd with the mint
    async fn infer_internal_payment(&self, invoice: Invoice) -> Result<bool> {
        // Check if this invoice is to be routed through any of the gateways as a final hop
        if let Some(route) = invoice.private_routes().last().take() {
            match (**route).0.last().take() {
                Some(last_hop) => {
                    // Check if the last hop for this invoice is a gateway registered with the mint
                    match self.fetch_registered_gateways().await {
                        Ok(gateways) => {
                            for gateway in gateways {
                                if gateway.node_pub_key == (*last_hop).src_node_id {
                                    return Ok(true);
                                }
                            }
                            Ok(false)
                        }
                        Err(_) => Ok(false),
                    }
                }
                None => Ok(false),
            }
        } else {
            Ok(false)
        }
    }

    // Application

    pub async fn fund_outgoing_ln_contract<R: RngCore + CryptoRng>(
        &self,
        invoice: Invoice,
        mut rng: R,
    ) -> Result<(ContractId, OutPoint)> {
        if self.infer_internal_payment(invoice.clone()).await? {
            // TODO: Instead of creating an outgoing contract to be executed by a gateway,
            // should we create an internal contract to be executed by the mint?
            unimplemented!("Direct settlement of internal payment not implemented yet");
        }
        // ... elided
    }

```

Q: Do we have any additional hints within a user client for detecting internal payments?
Q: Should the inference for an internal payment be done on attempt to fund an invoice and create an outgoing contract? Looks like this is the right place but I'm confirming
Q: What should we do to settle internal payments without delegating to the gateway?

**Comment by okjodom** [comment]
Discussed with @justinmoon, @dpc, @jkitman and @Kodylow in Discord general channel, concluding that we should prefer settlement of all payments should be delegated to the Gateways, including detection and settlement of internal payments.

This aligns with the overall design where :
- The mint does not facilitate any direct payments between users, and therefore does not collect fees / make money
- The gateway is a bridge and market maker between users within a mint
- The gateway is a bridge and market maker between users in a mint and the lightning network, including other mints over the network

**Comment by okjodom** [comment]
Closed by #461
